ports/nrf/main.c

```diff
@@ -110,6 +110,10 @@ int main(int argc, char **argv) {
 
 
 soft_reset:
+    #if defined(MICROPY_BOARD_EARLY_INIT)
+    MICROPY_BOARD_EARLY_INIT();
```

Comment:
On stm32 (and now renesas-ra) this call is outside (before) the soft reset loop.  If you want a callback here, it can be called `MICROPY_BOARD_TOP_SOFT_RESET_LOOP`

Reply:
Can we add all the missing macros ?  `MICROPY_BOARD_BEFORE_SOFT_RESET_LOOP`,  `MICROPY_BOARD_END_SOFT_RESET` etc.. ?

Reply:
yes

Reply:
Since there's not `boardctrl.h` in this port, should (empty) defaults be defined in `mpconfigport.h`, or use `#ifdef`s around each macro ?

Reply:
I think cleaner to define empty defaults in `mpconfigboard.h`, at least to start with.

Reply:
Actually I can't easily add those macros to `nrf` because they require the `boardctrl_state_t` and enum, and will have to add a `boardctrl.h`. FWIW, I think `boardctrl.h/c` should be common code.